core/slice.go

```diff
@@ -0,0 +1,484 @@
+package core
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sort"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/spruce-solutions/go-quai/common"
+	"github.com/spruce-solutions/go-quai/consensus"
+	"github.com/spruce-solutions/go-quai/core/rawdb"
+	"github.com/spruce-solutions/go-quai/core/types"
+	"github.com/spruce-solutions/go-quai/core/vm"
+	"github.com/spruce-solutions/go-quai/ethclient/quaiclient"
+	"github.com/spruce-solutions/go-quai/ethdb"
+	"github.com/spruce-solutions/go-quai/log"
+	"github.com/spruce-solutions/go-quai/params"
+)
+
+type Slice struct {
+	hc *HeaderChain
+
+	config *params.ChainConfig
+	engine consensus.Engine
+
+	quit chan struct{} // slice quit channel
+
+	domClient    *quaiclient.Client   // domClient is used to check if a given dominant block in the chain is canonical in dominant chain.
+	subClients   []*quaiclient.Client // subClinets is used to check is a coincident block is valid in the subordinate context
+	futureBlocks *lru.Cache
+
+	wg sync.WaitGroup // slice processing wait group for shutting down
+}
+
+func NewSlice(db ethdb.Database, chainConfig *params.ChainConfig, domClientUrl string, subClientUrls []string, engine consensus.Engine, vmConfig vm.Config) (*Slice, error) {
+	sl := &Slice{
+		config: chainConfig,
+		engine: engine,
+	}
+
+	futureBlocks, _ := lru.New(maxFutureBlocks)
+	sl.futureBlocks = futureBlocks
+
+	var err error
+	sl.hc, err = NewHeaderChain(db, engine, chainConfig, vmConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// only set the domClient if the chain is not prime
+	if types.QuaiNetworkContext != params.PRIME {
+		sl.domClient = MakeDomClient(domClientUrl)
+	}
+
+	sl.subClients = make([]*quaiclient.Client, 3)
```

Comment:
(nit) I'd prefer if we add a `params.MAX_SUBORDINATES` or `params.SUBORDINATE_DEGREE` or something that specifies the degree of the tree, instead of using the magic number `3`.

This is only a nit-pick today, but will become critical if we ever intend to support trees of varying degree later on.